master/buildbot/steps/artifact.py

```diff
@@ -35,7 +37,64 @@ def forceRebuild(build):
 
     return force_chain_rebuild or force_rebuild
 
-class FindPreviousSuccessfulBuild(ResumeBuild):
+class PreviousBuildStatus:
+    Forced = 1
+    Unmergeable = 2
+    NotFound = 3
+    Found = 4
+
+class FindPreviousSuccessBuildMixin():
+    @defer.inlineCallbacks
+    def _determinPreviousBuild(self, master, build, build_sourcestamps):
+        if forceRebuild(build):
+            defer.returnValue((PreviousBuildStatus.Forced, None))
+
+        mergeRequestFn = build.builder.getConfiguredMergeRequestsFn()
+
+        if mergeRequestFn == False:
+            defer.returnValue((PreviousBuildStatus.Unmergeable, None))
+
+        elif mergeRequestFn == True: # Default case, with no merge function assigned
+            prevBuildRequest = yield master.db.buildrequests \
+                .getBuildRequestBySourcestamps(buildername=build.builder.config.name,
+                                               sourcestamps=build_sourcestamps)
+            if prevBuildRequest:
+                defer.returnValue((PreviousBuildStatus.Found, prevBuildRequest))
+
+        else: # Custom merge function assigned
+            prevBuildRequests = yield master.db.buildrequests \
+                .getBuildRequestsBySourcestamps(buildername=build.builder.config.name,
+                                                sourcestamps = build_sourcestamps)
+            if len(prevBuildRequests) > 0 and len(build.requests) > 0:
+                req1 = build.requests[0]
+                for prevBuildRequest in prevBuildRequests:
+                    req2 = yield self._getBuildRequest(master, prevBuildRequest, req1.sources)
+                    if (mergeRequestFn(build.builder, req1, req2)):
+                        defer.returnValue((PreviousBuildStatus.Found, prevBuildRequest))
+
+        defer.returnValue((PreviousBuildStatus.NotFound, None))
+
+    def _updateMergedBuildRequests(self, master, build):
+        if len(build.requests) > 1:
+            yield master.db.buildrequests.updateMergedBuildRequest(build.requests)
+
+    # A more light weight function for making a build request, it uses as much of the information we already have to
+    # make a build request - so we only need to fetch the properties
+    @defer.inlineCallbacks
+    def _getBuildRequest(self, master, brdict, sources):
```

Comment:
`n+1` selects, I'll try to batch select them outside the loop and then split the results inside here